File: kotlin/matching-brackets/src/main/kotlin/MatchingBrackets.kt
object MatchingBrackets {

    private val brackets = mapOf(
        '{' to '}',
        '(' to ')',
        '[' to ']',
    )

    fun isValid(input: String) = isValid(emptyList(), input)

    private tailrec fun isValid(stack: List<Char>, input: String): Boolean {
        if (input.isBlank()) {
            return stack.isEmpty()
        }

        val head = input.first()
        val tail = input.drop(1)
        val expected = stack.firstOrNull()?.let { brackets[it] }

        return when (head) {
            in brackets.keys -> isValid(listOf(head) + stack, tail)
            in brackets.values -> if (head == expected) isValid(stack.drop(1), tail) else false
            else -> isValid(stack, tail)
        }
    }
}

File: scala/matching-brackets/src/main/scala/MatchingBrackets.scala
import scala.annotation.tailrec

object MatchingBrackets {
  private val brackets = Map(('(', ')'), ('[', ']'), ('{', '}'))

  def isPaired(brackets: String): Boolean = isPaired(List(), brackets)

  @tailrec
  private def isPaired(stack: List[Char], input: String): Boolean = {
    val expected = stack.headOption.flatMap(brackets.get)

    input.headOption match {
      case None => stack.isEmpty
      case Some(head) if brackets.keySet contains head => isPaired(head :: stack, input.tail)
      case Some(head) if brackets.values.toList contains head =>
        if (expected.contains(head)) isPaired(stack.tail, input.tail) else false
      case _ => isPaired(stack, input.tail)
    }
  }
}
